cli/mender-artifact/copy.go

```diff
@@ -0,0 +1,256 @@
+)
+
+func Cat(c *cli.Context) (err error) {
+	if c.NArg() != 1 {
+		return cli.NewExitError(fmt.Sprintf("Got %d arguments, wants one", c.NArg()), 1)
+	}
+	r, err := NewPartitionReader(c.Args().First(), c.String("key"))
+	if err != nil {
+		return cli.NewExitError(fmt.Sprintf("failed to open the partition reader: err: %v", err), 1)
+	}
+	var w io.WriteCloser = os.Stdout
+	if _, err = io.Copy(w, r); err != nil {
+		return cli.NewExitError(fmt.Sprintf("failed to copy from: %s to stdout: err: %v", c.Args().First(), err), 1)
+	}
+	return r.Close()
+}
+
+func Copy(c *cli.Context) (err error) {
+	var r io.ReadCloser
+	var w io.WriteCloser
+
+	var repack bool
+	if err := doCopy(c, &repack, &r, &w); err != nil {
+		return err
+	}
+
+	_, err = io.Copy(w, r)
+	if err != nil {
+		return cli.NewExitError(err.Error(), 1)
+	}
+
+	// Only repack in case of writing to an image, a read will not change the image,
+	// and thus we spare a fair few seconds not repacking it.
+	if repack {
+		wpc, ok := w.(PartitionPacker)
+		if !ok {
+			return cli.NewExitError("critical implementation error. Image cannot be repacked", 1)
+		}
+		if err = wpc.Repack(); err != nil {
+			return cli.NewExitError(fmt.Sprintf("failed to repack image: %v", err), 1)
+		}
+
+	}
+	if err = w.Close(); err != nil {
+		return err
+	}
+	return r.Close()
+}
+
+func doCopy(c *cli.Context, repack *bool, r *io.ReadCloser,
+	w *io.WriteCloser) error {
+
+	var err error
+
+	switch parseCLIOptions(c) {
+	case copyin:
+		*r, err = os.OpenFile(c.Args().First(), os.O_CREATE|os.O_RDWR, 0655)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		*w, err = NewPartitionWritePacker(c.Args().Get(1), c.String("key"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		*repack = true
+	case copyinstdin:
+		*r = os.Stdin
+		*w, err = NewPartitionWritePacker(c.Args().First(), c.String("key"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		*repack = true
+	case copyout:
+		*r, err = NewPartitionReader(c.Args().First(), c.String("key"))
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+		*w, err = os.OpenFile(c.Args().Get(1), os.O_CREATE|os.O_RDWR, 0655)
+		if err != nil {
+			return cli.NewExitError(fmt.Sprintf("%v", err), 1)
+		}
+	case parseError:
+		return cli.NewExitError(fmt.Sprintln("no artifact or image provided"), 1)
+	case argerror:
+		return cli.NewExitError(fmt.Sprintf("got %d arguments, wants two", c.NArg()), 1)
+	default:
+		return cli.NewExitError("critical error", 1)
+	}
+	return nil
+}
+
+func Append(c *cli.Context) error {
+	var rwcp PartitionReadWriteClosePacker
+	var appendint = 3
+	var readindex = 0
+	var err error
+
+	if c.NArg() != 2 {
+		return cli.NewExitError("wrong number of input arguments. Two required", 1)
+	}
+
+	for i, arg := range c.Args() {
+		if arg == "-" {
+			appendint = i
+			break
+		}
+	}
+
+	isImg := regexp.MustCompile(`\.(mender|sdimg)`)
```

Comment:
Could you add `uefiimg` to this while you're at it? This will become relevant very soon, and the code for producing such an image is already in meta-mender/master.

Reply:
No need. I’ve decided to remove the append functionality.